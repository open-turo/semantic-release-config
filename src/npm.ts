import config from "./";

// We have a set of plugins that ideally should run after every other plugin to guarantee that things like publishing to NPM
// already happen before these plugins run
const pluginsThatGoAtTheEnd = new Set([
  "@semantic-release/exec",
  "semantic-release-fotingo",
]);

/**
 * Given a semantic release plugin config, return the plugin name
 * @param p Plugin config
 * @returns Plugin name
 */
const getPluginName = (p: typeof config.plugins[number]) => {
  if (typeof p === "string") {
    return p;
  }
  return p[0];
};

/**
 * Semantic release configuration for NPM projects. It extends the default configuration by publishing to NPM
 * and creating a Pull Request with version updates in files like package.json
 */
export = {
  plugins: [
    ...config.plugins.filter(
      (p) => !pluginsThatGoAtTheEnd.has(getPluginName(p))
    ),
    [
      "@semantic-release/npm",
      {
        tarballDir: "pack",
      },
    ],
    [
      "@semantic-release/git",
      {
        assets: ["package.json", "package-lock.json", "README.md", "yarn.lock"],
        message:
          "ci(release): ${nextRelease.version} <% nextRelease.channel !== 'next' ? print('[skip ci]') : print('') %>\n\n${nextRelease.notes}",
      },
    ],
    ...config.plugins.filter((p) =>
      pluginsThatGoAtTheEnd.has(getPluginName(p))
    ),
  ],
};
