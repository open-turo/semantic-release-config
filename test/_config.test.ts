import {
  createPreset,
  semanticReleaseGit,
  SemanticReleasePlugin,
} from "~/_config";
import npm from "~/npm";
import template = require("lodash.template");

describe("config", () => {
  describe("createPreset", () => {
    let preset: { plugins: SemanticReleasePlugin[] };
    beforeEach(() => {
      preset = createPreset(["a", "b"]);
    });
    test("creates a preset including the default config", () => {
      expect(preset).toMatchSnapshot();
    });

    test("@semantic-release/exec and semantic-release-fotingo are the last plugins", () => {
      expect(preset.plugins[npm.plugins.length - 2]).toBe(
        "semantic-release-fotingo"
      );
      expect(preset.plugins[npm.plugins.length - 1][0]).toBe(
        "@semantic-release/exec"
      );
    });
  });

  describe("semanticReleaseGit", () => {
    test.each(["next", undefined])(
      "semantic-release/@git generates the right commit message for channel %s",
      (channel) => {
        const assets = ["a"];
        const gitPlugin = semanticReleaseGit(assets);
        const pluginConfig = ((gitPlugin && gitPlugin[1]) || {
          assets: [],
          message: "",
        }) as {
          assets: [];
          message: string;
        };
        expect(pluginConfig.assets).toBe(assets);
        expect(
          template(pluginConfig.message || "")({
            nextRelease: {
              version: "test",
              channel: channel,
              notes: "This is a test release",
            },
          })
        ).toMatchSnapshot();
      }
    );
  });
});
