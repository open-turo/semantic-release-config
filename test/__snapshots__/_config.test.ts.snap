// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config createPreset creates a preset including the default config 1`] = `
{
  "branches": [
    "main",
    "+([0-9])?(.{+([0-9]),x}).x",
    {
      "channel": "next",
      "name": "(f|b|c)/*",
      "prerelease": "beta-\${(/^[a-zA-Z]+-[0-9]+/.exec(name.substr(2)) || [name.replace(/[_/.]/g, '-')])[0]}-\${Date.now()}",
    },
  ],
  "plugins": [
    "@semantic-release/commit-analyzer",
    [
      "@semantic-release/release-notes-generator",
      {
        "parserOpts": {
          "noteKeywords": [
            "BREAKING CHANGE",
            "BREAKING CHANGES",
            "BREAKING",
          ],
        },
        "preset": "conventionalcommits",
        "presetConfig": {
          "types": [
            {
              "hidden": false,
              "section": "Build System",
              "type": "build",
            },
            {
              "hidden": false,
              "section": "Miscellaneous",
              "type": "chore",
            },
            {
              "hidden": false,
              "section": "Continuous Integration",
              "type": "ci",
            },
            {
              "hidden": false,
              "section": "Documentation",
              "type": "docs",
            },
            {
              "hidden": false,
              "section": "Features",
              "type": "feat",
            },
            {
              "hidden": false,
              "section": "Bug Fixes",
              "type": "fix",
            },
            {
              "hidden": false,
              "section": "Performance Improvements",
              "type": "perf",
            },
            {
              "hidden": false,
              "section": "Code Refactoring",
              "type": "refactor",
            },
            {
              "hidden": false,
              "section": "Styles",
              "type": "style",
            },
            {
              "hidden": false,
              "section": "Tests",
              "type": "test",
            },
          ],
        },
        "writerOpts": {
          "commitsSort": [
            "subject",
            "scope",
          ],
        },
      },
    ],
    "@semantic-release/github",
    "a",
    "b",
    "semantic-release-fotingo",
    [
      "@semantic-release/exec",
      {
        "failCmd": "[ -x ./script/semantic-release-fail ] && ./script/semantic-release-fail '\${nextRelease.version}' || true",
        "publishCmd": "[ -x ./script/semantic-release-publish ] && ./script/semantic-release-publish '\${nextRelease.version}' || true",
        "successCmd": "[ -x ./script/semantic-release-success ] && ./script/semantic-release-success '\${nextRelease.version}' || true",
      },
    ],
  ],
}
`;

exports[`config semanticReleaseGit semantic-release/@git generates the right commit message for channel next 1`] = `
"ci(release): test 

This is a test release"
`;

exports[`config semanticReleaseGit semantic-release/@git generates the right commit message for channel undefined 1`] = `
"ci(release): test [skip ci]

This is a test release"
`;
